helix-lsp/src/snippet.rs

```diff
@@ -0,0 +1,507 @@
+        None => return Transaction::new(doc),
+    };
+
+    let newline_with_offset = format!(
+        "{line_ending}{blank:width$}",
+        width = edit.range.start.character as usize,
+        blank = ""
+    );
+
+    let mut insert = String::new();
+    let mut offset = (primary_cursor as i128 + start_offset) as usize;
+    let mut tabstops: Vec<(usize, Range)> = Vec::new();
+
+    for element in snippet.elements {
+        match element {
+            Text(text) => {
+                // small optimization to avoid calling replace when it's unnecessary
+                let text = if text.contains('\n') {
+                    Cow::Owned(text.replace('\n', &newline_with_offset))
+                } else {
+                    Cow::Borrowed(text)
+                };
+                offset += text.chars().count();
+                insert.push_str(&text);
+            }
+            Variable {
+                name: _name,
+                regex: None,
+                r#default,
+            } => {
+                // TODO: variables. For now, fall back to the default, which defaults to "".
+                let text = r#default.unwrap_or_default();
+                offset += text.chars().count();
+                insert.push_str(text);
+            }
+            Tabstop { tabstop } => {
+                tabstops.push((tabstop, Range::point(offset)));
+            }
+            Placeholder { tabstop, value } => match value.as_ref() {
+                // https://doc.rust-lang.org/beta/unstable-book/language-features/box-patterns.html
+                // would make this a bit nicer
+                Text(text) => {
+                    if include_placeholer {
+                        let len_chars = text.chars().count();
+                        tabstops.push((tabstop, Range::new(offset, offset + len_chars + 1)));
+                        offset += len_chars;
+                        insert.push_str(text);
+                    } else {
+                        tabstops.push((tabstop, Range::point(offset)));
+                    }
+                }
+                other => {
+                    log::error!(
+                        "Discarding snippet: generating a transaction for placeholder contents {:?} is unimplemented.",
+                        other
+                    );
+                    return Transaction::new(doc);
+                }
+            },
+            other => {
+                log::error!(
+                    "Discarding snippet: generating a transaction for {:?} is unimplemented.",
+                    other
+                );
+                return Transaction::new(doc);
+            }
+        }
+    }
+
+    let transaction = Transaction::change_by_selection(doc, selection, |range| {
+        let cursor = range.cursor(text);
+        (
+            (cursor as i128 + start_offset) as usize,
+            (cursor as i128 + end_offset) as usize,
+            Some(insert.clone().into()),
```

Comment:
this code can generate out of bounds edits and crashes (see #5728).

At first I considered refactoring this function to reuse`generate_transaction_from_completion_edit` but I think that is actually not the right approach. When variables are used then the completion at each cursor must be different. 

For now I am just leaving this here as a note so remember that this needs to be fixed up later after #5728  is merged.

Reply:
I think using generate_transaction_from_completion_edit logic is actually fine if generation of the change text is delayed until we know the cursor position (and so is re-done for each cursor).